extensions/devservices/common/src/main/java/io/quarkus/devservices/common/ContainerLocator.java

```diff
@@ -0,0 +1,58 @@
+package io.quarkus.devservices.common;
+
+import java.util.Arrays;
+import java.util.Optional;
+
+import org.jboss.logging.Logger;
+import org.testcontainers.DockerClientFactory;
+
+import com.github.dockerjava.api.model.Container;
+import com.github.dockerjava.api.model.ContainerPort;
+
+import io.quarkus.runtime.LaunchMode;
+
+public class ContainerLocator {
+
+    private static final Logger log = Logger.getLogger(ContainerLocator.class);
+
+    private final String devServiceLabel;
+    private final int port;
+
+    public ContainerLocator(String devServiceLabel, int port) {
+        this.devServiceLabel = devServiceLabel;
+        this.port = port;
+    }
+
+    private Optional<Container> lookup(String expectedLabelValue) {
+        return DockerClientFactory.lazyClient().listContainersCmd().exec().stream()
+                .filter(container -> expectedLabelValue.equals(container.getLabels().get(devServiceLabel)))
+                .findAny();
+    }
+
+    private Optional<ContainerPort> getMappedPort(Container container, int port) {
+        return Arrays.stream(container.getPorts())
+                .filter(containerPort -> containerPort.getPrivatePort() != null)
+                .filter(containerPort -> containerPort.getPublicPort() != null)
+                .filter(containerPort -> containerPort.getPrivatePort() == port)
```

Comment:
Minor: The 3 filters can be merged in a single one.

Reply:
@cescoffier Yes, I did consider it, but I didn't want to make this line too long - as right now it's easier to read.
Actually now, when I saw this code again with a fresh head - I think it could be extracted to a predicate, so we will a single filter - which won't be too long at the same time.